[vcpkg] vcpkg_configure_meson: cross-compile fixes for Android

jwtowner opened the pull request:
This PR includes two fixes to the meson cross-compile config.

The first change fixes the regex used to partition linker flag options so that it supports double-hyphen prefixes. Without this change, passing a linker flag such as "--target=armv7a-linux-androideabi21" will end up being passed incorrectly as "- -target=armv7a-linux-androideabi21".

The second change ensures that `vcpkg_internal_meson_generate_cross_file` always generates a cross file when the target platform is either Android or iOS, even if the host and target CPU architectures are the same. For example, when cross compiling harfbuzz:x64-android on a Windows 64-bit (x86-64) host, `vcpkg_internal_meson_generate_cross_file` normally does not generate a cross file because the host and target CPU architectures are equal, and this in turn causes Meson to attempt to run the generated x64-android sanity checking binaries as if they were Windows binaries, which causes the configuration to fail altogether.

One question, is there a better way to check if `VCPKG_CMAKE_SYSTEM_NAME` does not match the host system, and in a way that will be future proofed for other cross-compile target systems other than Android and iOS?

This PR, in tandem with PR #15605, allows for harfbuzz (and potentially other ports which also use meson) to be built for arm-android, arm64-android, x86-android and x64-android. Despite harfbuzz:*-android also relying on #15605, this PR should be safe to merge independently of it.

jwtowner committed:
[vcpkg] vcpkg_configure_meson: cross-compile fixes

This change includes two fixes to the meson cross-compile config.

The first change fixes the regex used to partition linker flag
options so that it supports double-hyphen prefixes.

The second change ensures that the cross file is always generated
when the target platform is either Android or iOS, even if the
host and target CPU architectures are the same.

Neumann-A commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -128,7 +128,7 @@ function(vcpkg_internal_meson_generate_native_file_config _config) #https://meso
         set(LINKER_FLAGS_${_config} "${VCPKG_DETECTED_CMAKE_STATIC_LINKER_FLAGS_${_config}}")
     endif()
-    string(REGEX REPLACE "( |^)(-|/)" ";\\2" LINKER_FLAGS_${_config} "${LINKER_FLAGS_${_config}}")
+    string(REGEX REPLACE "( |^)(--?|/)" ";\\2" LINKER_FLAGS_${_config} "${LINKER_FLAGS_${_config}}")
```
Use `vcpkg_internal_meson_convert_compiler_flags_to_list` instead

jwtowner replied:
Fixed in 8242f9e

Neumann-A commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -316,7 +316,7 @@ function(vcpkg_internal_meson_generate_cross_file_config _config) #https://meson
         set(LINKER_FLAGS_${_config} "${VCPKG_DETECTED_CMAKE_STATIC_LINKER_FLAGS_${_config}}")
     endif()
-    string(REGEX REPLACE "( |^)(-|/)" ";\\2" LINKER_FLAGS_${_config} "${LINKER_FLAGS_${_config}}")
+    string(REGEX REPLACE "( |^)(--?|/)" ";\\2" LINKER_FLAGS_${_config} "${LINKER_FLAGS_${_config}}")
```
same as above

jwtowner replied:
Fixed in 8242f9e

jwtowner committed:
Use helper function instead of regex replace

Use vcpkg_internal_meson_convert_compiler_flags_to_list to partition
compiler flags into a list, instead of string(REGEX REPLACE ...).

ras0219-msft commented:
Awesome, thanks for the fixes!
